### PR TITLE
ROC-2475: Add UI to view of offer to settle the claim

### PR DESCRIPTION
**Pull request by comanirudha**
### JIRA link ###
https://tools.hmcts.net/jira/browse/ROC-2475

### Change description ###

Claimant view of offer to settle the claim

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/features/offer/routes/defendant-response.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    form: form,
+    claim : res.locals.user.claim,
+    offer: getOffer(res.locals.user.claim)
+  })
+}
+
+export default express.Router()
+  .get(
+    Paths.defendantResponsePage.uri,
+    OfferGuard.requestHandler,
```
Comment (geek-soft) [inline src/main/features/offer/routes/defendant-response.ts]: It does not make sense to have it here since guard redirects to conformation page. See below logic extracted from the guard.

```
logger.warn(`Claim ${claim.claimNumber} not eligible for an offer - redirecting to confirmation page`)
res.redirect(Paths.confirmationPage.evaluateUri({ externalId: user.claim.externalId }))
```

Please also use `GuardFactory.create` in the guard so that we have single class to mock and common mechanism of writing guards.

File: src/main/features/offer/routes/defendant-response.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import { DefendantResponse } from 'offer/form/models/defendantResponse'
+import { ErrorHandling } from 'common/errorHandling'
+import User from 'idam/user'
+import { OfferGuard } from 'offer/guards/offerGuard'
+
+function getOffer (claim: any): any {
+  if (!claim.settlement || !claim.settlement.partyStatements) {
+    return undefined
+  }
+  return claim.settlement.partyStatements[0].offer
```
Comment (geek-soft) [inline src/main/features/offer/routes/defendant-response.ts]: This should return last offer not the first one since the page will be reused for responding to more then one defendant offer. Implement something similar to `uk.gov.hmcts.cmc.claimstore.models.offers.Settlement#getLastStatement`

File: src/main/features/offer/views/defendant-response.njk [hunk 3]
```diff
@@ -0,0 +1,69 @@
+{% extends "layout.njk" %}
+{% from "form.njk" import csrfProtection, submitButton, errorSummary, radioGroup %}
+{% from "date.njk" import dateInput %}
+
+{% set title = t('Money claim') %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+      <h1 class="heading-large form-hint">{{ t("Defendant's response") }} </h1>
```
Comment (geek-soft) [inline src/main/features/offer/views/defendant-response.njk]: Please use single quotes and correct apostrophe char.

File: src/main/features/offer/views/defendant-response.njk [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      <form novalidate method="post">
+        {{ csrfProtection(csrf) }}
+        <div class="description-block">
+          <h2 class="heading-large"> 
+          {{ t('%s’s offer',{
+              postProcess: 'sprintf',
+              sprintf: [claim.claimData.defendant.name]
+            }) }}</h2>
+        </div>
+        <table class="claim-dashboard">
```
Comment (geek-soft) [inline src/main/features/offer/views/defendant-response.njk]: That does not make sense to have it here. Please remove or reimplement for an offer table.

File: src/main/features/offer/views/defendant-response.njk [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+          <tr class="claims default mobile-table">
```
Comment (geek-soft) [inline src/main/features/offer/views/defendant-response.njk]: The same here, why `claims` class is used here?

Comment (geek-soft) [reply src/main/features/offer/views/defendant-response.njk]: What `default` class is doing here for us? Is it declared anywhere?

File: src/main/features/offer/views/defendant-response.njk [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+              <td>{{ offer.completionDate | date }}</td>
+            </tr>
+        </table>
+
+        <h2 class="heading-medium"> {{ t('Do you accept the offer?') }} </h2>
+
+        {{
+        radioGroup(
+          form = form,
+          label = 'Yes',
```
Comment (geek-soft) [inline src/main/features/offer/views/defendant-response.njk]: That does not look right. That is the label for fieldset

File: src/test/features/offer/routes/defendant-response.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            await request(app)
+              .post(defendantResponsePage)
+              .set('Cookie', `${cookieName}=ABC`)
+              .send({})
+              .expect(res => expect(res).to.be.serverError.withText('Error'))
+          })
+        })
+
+        context('when form is valid', async () => {
+          it('should display defendantResponsePage', async () => {
```
Comment (geek-soft) [inline src/test/features/offer/routes/defendant-response.ts]: We tend to say `should render page`. No need to be specific since that make maintenance harder in case page is renamed.

File: src/main/features/offer/paths.ts [hunk 8]
```diff
@@ -6,4 +6,5 @@ export class Paths {
   static readonly offerPage = new RoutablePath(`${offerPath}/your-offer`)
   static readonly offerConfirmationPage = new RoutablePath(`${offerPath}/offer-confirmation`)
   static readonly settleOutOfCourtPage = new RoutablePath(`${offerPath}/settle-out-of-court`)
+  static readonly defendantResponsePage = new RoutablePath(`${offerPath}/defendant-response`)
```
[the remark this answers is not in the document]
Comment (geek-soft) [reply src/main/features/offer/paths.ts]: What about `/offer/summary` or `/offer/response`?

[the remark this answers is not in the document]
Comment (comanirudha) [reply src/main/features/offer/paths.ts]: /offer/response sounds right, I will change it.

File: src/main/app/claims/models/partyStatement.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
+
+import { Serializable } from 'models/serializable'
+import { Offer } from 'claims/models/offer'
+
+export class PartyStatement implements Serializable<PartyStatement> {
+  type: string
+  madeBy: string
+  offer: Offer
```
Comment (rlewan) [inline src/main/app/claims/models/partyStatement.ts]: This field is optional on the backend side, should be optional here as well.

File: src/main/features/offer/form/models/defendantResponse.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import { IsDefined, IsIn } from 'class-validator'
+import { Serializable } from 'models/serializable'
+
+export class ValidationErrors {
+  static readonly OPTION_REQUIRED: string = 'Choose option: yes or no or make an offer'
+}
+
+export class DefendantResponse implements Serializable<DefendantResponse> {
+  @IsDefined({ message: ValidationErrors.OPTION_REQUIRED })
+  @IsIn(['yes', 'no', 'makeAnCounterOffer'], { message: ValidationErrors.OPTION_REQUIRED })
```
Comment (rlewan) [inline src/main/features/offer/form/models/defendantResponse.ts]: These values should be defined in a constants class , e.g. like in [`EvidenceType`](https://github.com/hmcts/cmc-citizen-frontend/blob/master/src/main/features/response/form/models/evidenceType.ts). This constants class should be also exported as a nunjucks global, see [here](https://github.com/hmcts/cmc-citizen-frontend/blob/master/src/main/modules/nunjucks/index.ts#L78) for details.

File: src/main/features/offer/routes/defendant-response.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+import { Paths } from 'offer/paths'
+import { Form } from 'forms/form'
+import { DefendantResponse } from 'offer/form/models/defendantResponse'
+import { ErrorHandling } from 'common/errorHandling'
+import User from 'idam/user'
+import { OfferGuard } from 'offer/guards/offerGuard'
+
+function getOffer (claim: any): any {
+  if (!claim.settlement || !claim.settlement.partyStatements) {
+    return undefined
```
Comment (rlewan) [inline src/main/features/offer/routes/defendant-response.ts]: You are returning `undefined` here, but assuming that the offer will be present on the page. Please either make an assertion here that an offer is present when this route is accessed or add logic to the page which correctly handles missing offer. I would lean towards the former.

File: src/main/app/claims/models/offer.ts [hunk 12]
```diff
@@ -7,7 +7,7 @@ export class Offer implements Serializable<Offer> {
   content: string
   completionDate: Moment
 
-  constructor (content: string, completionDate: Moment) {
+  constructor (content?: string, completionDate?: Moment) {
```
[the remark this answers is not in the document]
Comment (comanirudha) [reply src/main/app/claims/models/offer.ts]: The current pattern is to use X(). deserialise, may be we need a overall code refactoring which can be done as a separate story.

[the remark this answers is not in the document]
Comment (geek-soft) [reply src/main/app/claims/models/offer.ts]: Definitely not for now, I just pointed out the side effect of `new X().deserialise` pattern which I do not like.

File: src/main/features/offer/form/models/defendantResponse.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+import { Serializable } from 'models/serializable'
+import { StatementType } from './statementType'
+export class ValidationErrors {
+  static readonly OPTION_REQUIRED: string = 'Choose option: yes or no or make an offer'
+}
+
+export class DefendantResponse implements Serializable<DefendantResponse> {
+  @IsDefined({ message: ValidationErrors.OPTION_REQUIRED })
+  @IsIn([StatementType.ACCEPTATION.value, StatementType.REJECTION.value, StatementType.OFFER.value], { message: ValidationErrors.OPTION_REQUIRED })
+  option?: string
```
Comment (geek-soft) [inline src/main/features/offer/form/models/defendantResponse.ts]: Please add validation unit test similar to `features/response/form/models/rejectPartOfClaim.ts` for example.

File: src/main/features/offer/routes/defendant-response.ts (hunk 1 above)
Comment (geek-soft) [reply src/main/features/offer/routes/defendant-response.ts]: This has not been addressed yet!!!

File: src/main/features/offer/routes/response.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+import { Paths } from 'offer/paths'
+import { Form } from 'forms/form'
+import { DefendantResponse } from 'offer/form/models/defendantResponse'
+import { ErrorHandling } from 'common/errorHandling'
+import User from 'idam/user'
+import { OfferGuard } from 'offer/guards/offerGuard'
+
+function getPartyStatement (claim: any): any[] {
+  if (!claim.settlement || !claim.settlement.partyStatements) {
+    throw Error('Something went wrong, No offer present')
```
Comment (geek-soft) [inline src/main/features/offer/routes/response.ts]: This is not desired solution. Having this would mean that when citizen hits the page URI at a time there is no offer they will get 500 error. We wanna handle this scenario more gracefully so the possible solutions are:

- having business error page saying offer has not been made by defendant yet or more generic one saying an action is not available at this stage (preferred)
- redirect silently to dashboard without error message

File: src/main/features/offer/routes/response.ts [hunk 15]
```diff
@@ -0,0 +1,56 @@
+import * as express from 'express'
+import { FormValidator } from 'forms/validation/formValidator'
+import { Paths } from 'offer/paths'
+import { Form } from 'forms/form'
+import { DefendantResponse } from 'offer/form/models/defendantResponse'
+import { ErrorHandling } from 'common/errorHandling'
+import User from 'idam/user'
+import { OfferGuard } from 'offer/guards/offerGuard'
+
+function getPartyStatement (claim: any): any[] {
```
Comment (geek-soft) [inline src/main/features/offer/routes/response.ts]: Please declare return type correctly. We aim to avoid any. In your case it is `PartyStatement[]`

File: src/main/app/claims/models/partyStatement.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
+
+import { Serializable } from 'models/serializable'
+import { Offer } from 'claims/models/offer'
+
+export class PartyStatement implements Serializable<PartyStatement> {
+  type?: string
+  madeBy?: string
+  offer?: Offer
```
Comment (rlewan) [inline src/main/app/claims/models/partyStatement.ts]: I my previous comment I meant that only `offer` is an optional field. Other two should be mandatory.

File: src/main/features/offer/views/response.njk [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+          name = 'option',
+          options = [
+            {
+              bold: false,
+              label: 'Yes',
+              hint: t('You can make a legal agreement between you and %s.',
+                { postProcess: 'sprintf',
+                sprintf:[claim.claimData.defendant.name]
+                }),
+              value: 'StatementType.ACCEPTATION.value',
```
Comment (rlewan) [inline src/main/features/offer/views/response.njk]: You are passing the `'StatementType.ACCEPTATION.value'` string here, not accessing `value` property of `StatementType.ACCEPTATION`. Same in other places below.

File: src/test/http-mocks/claim-store.ts [hunk 18]
```diff
@@ -66,6 +66,10 @@ export const sampleClaimObj = {
     defendantDateOfBirth: '1990-11-01',
     paidAmount: 2,
     paymentOption: 'IMMEDIATELY'
+  },
+  settlement: {
+    partyStatements:
+        [ { type: 'OFFER', madeBy: 'DEFENDANT' , offer: { content: 'offer text', completionDate: '2017-08-08'}}]
```
Comment (rlewan) [inline src/test/http-mocks/claim-store.ts]: Can we use the new constants here as well?

File: src/main/features/offer/form/models/defendantResponse.ts (hunk 13 above)
Comment (comanirudha) [reply src/main/features/offer/form/models/defendantResponse.ts]: we already have some router tests which test validation.

File: src/main/features/offer/routes/response.ts (hunk 14 above)
Comment (comanirudha) [reply src/main/features/offer/routes/response.ts]: the use can only see this page if they have offer, if they don't the link is usually disabled. This if only safeguards against directly entering url in the browser.

File: src/main/features/offer/form/models/defendantResponse.ts (hunk 13 above)
Comment (comanirudha) [reply src/main/features/offer/form/models/defendantResponse.ts]: done

**Comment by timja-kainos** [comment]
conflicts

File: src/main/app/claims/models/partyStatement.ts [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+    this.type = type
+    this.madeBy = madeBy
+    this.offer = offer
+  }
+
+  deserialize (input: any): PartyStatement {
+    if (input) {
+      this.type = input.type
+      this.madeBy = input.madeBy
+      this.offer = new Offer().deserialize(input.offer)
```
Comment (geek-soft) [inline src/main/app/claims/models/partyStatement.ts]: If offer is optional I think we should have a condition here so that when undefined was returned from backend we have `PartyStatement.undefined `. I think about something like:

```
if (input.offer) {
  this.offer = new Offer().deserialize(input.offer)
}
```

File: src/main/features/offer/form/models/defendantResponse.ts (hunk 13 above)
Comment (geek-soft) [reply src/main/features/offer/form/models/defendantResponse.ts]: Just to clarify route test is not a place to test validation. I roughly test that validation is enabled but not that validation works fine. In route tests we test one scenario but here we have 3 possible options and we need to test all are handled properly.

File: src/main/features/offer/form/models/offer.ts [hunk 20]
```diff
@@ -17,24 +17,27 @@ export class Offer implements Serializable<Offer> {
...
   @IsDefined({ message: ValidationErrors.DATE_REQUIRED })
   @IsValidLocalDate({ message: ValidationErrors.DATE_NOT_VALID })
   @IsFutureDate({ message: ValidationErrors.FUTURE_DATE })
-  completionDate?: LocalDate
+  completionDate: LocalDate
 
-  constructor (offerText?: string, completionDate?: LocalDate) {
-    this.offerText = offerText
-    this.completionDate = completionDate
+  static create (offerText: string, completionDate: LocalDate) {
```
Comment (geek-soft) [inline src/main/features/offer/form/models/offer.ts]: What is the purpose of this? Why did you introduce it over regular constructor? If there is no significant advantage of this I would suggest reverting back to constructor as this is the pattern / expected way of instantiating classes.

File: src/main/features/offer/form/models/offer.ts [hunk 21]
```diff
@@ -17,24 +17,27 @@ export class Offer implements Serializable<Offer> {
...
+    const offer = new Offer()
+    if (offerText || completionDate) {
```
Comment (geek-soft) [inline src/main/features/offer/form/models/offer.ts]: What is the purpose of this? Populating fields if any of two arguments is provided looks wired at least. This also in conflict with the facts that both fields are marked as mandatory but here it is possible to instantiate only one field.

File: src/main/features/offer/routes/response.ts [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+import User from 'idam/user'
+import { OfferGuard } from 'offer/guards/offerGuard'
+import { StatementType } from 'offer/form/models/statementType'
+import { MadeBy } from 'offer/form/models/madeBy'
+
+function redirectToDashBoardPage (res: express.Response) {
+  const user: User = res.locals.user
+  res.redirect(DashboardPaths.claimantPage.evaluateUri({ externalId: user.claim.externalId }))
+}
+function getPartyStatement (claim: any, res: express.Response): PartyStatement[] {
```
Comment (geek-soft) [inline src/main/features/offer/routes/response.ts]: Correct type, claim is of type of Claim, not any. Lack of types is error prone.

File: src/main/features/offer/routes/response.ts [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+    redirectToDashBoardPage(res)
+  }
+
+  return claim.settlement.partyStatements.map(partyStatement => {
+    if (partyStatement.type === StatementType.OFFER.value && partyStatement.madeBy === MadeBy.DEFENDANT.value) {
+      return partyStatement
+    }
+  })
+}
+function getOffer (claim: any, res: express.Response): Offer {
```
Comment (geek-soft) [inline src/main/features/offer/routes/response.ts]: Correct type, remove `res` argument.

File: src/main/features/offer/routes/response.ts [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+import { StatementType } from 'offer/form/models/statementType'
+import { MadeBy } from 'offer/form/models/madeBy'
+
+function redirectToDashBoardPage (res: express.Response) {
+  const user: User = res.locals.user
+  res.redirect(DashboardPaths.claimantPage.evaluateUri({ externalId: user.claim.externalId }))
+}
+function getPartyStatement (claim: any, res: express.Response): PartyStatement[] {
+  if (!claim.settlement || !claim.settlement.partyStatements) {
+    redirectToDashBoardPage(res)
```
Comment (geek-soft) [inline src/main/features/offer/routes/response.ts]: This is side effect. A `getPartyStatement` should only return value, not do redirects. It violates CQRS principle.

Do that logic in the place which retrieved the value from `getPartyStatement`. If undefined you can make redirect. Also guard could be possible solution as probably no point even starting processing in that router when settlement is not available.

File: src/main/features/offer/routes/response.ts (hunk 23 above)
Comment (geek-soft) [reply src/main/features/offer/routes/response.ts]: Not a valid place to make redirect. See my comment above.

File: src/test/features/offer/form/models/defendantResponse.ts [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+      expect(new DefendantResponse().deserialize({ option: StatementType.ACCEPTATION.value})).to.deep.equal(new DefendantResponse(StatementType.ACCEPTATION.value))
+    })
+  })
+
+  describe('validation', () => {
+    const validator: Validator = new Validator()
+
+    describe('should reject when', () => {
+      it('undefined option', () => {
+        const errors = validator.validateSync(new DefendantResponse())
```
Comment (geek-soft) [inline src/test/features/offer/form/models/defendantResponse.ts]: Please pass undefined explicitly, in case default behaviour of the constructor is changed.

Plase also test that it is rejected when non valid option is passed, e.g. `new DefendantResponse('unknown-value')`

File: src/test/features/offer/form/models/offer.ts [hunk 26]
```diff
@@ -69,10 +69,10 @@ describe('Offer', () => {
 
     describe('should accept when', () => {
       it('offer text and future date', () => {
-        const futureDate = moment().add(10, 'days')
-        const date = new LocalDate(futureDate.year(), futureDate.month(), futureDate.day())
-        const errors = validator.validateSync(new Offer('offer text', date))
-        expect(errors.length).to.equal(1)
+        const futureDate = moment().add(10, 'month')
```
Comment (geek-soft) [inline src/test/features/offer/form/models/offer.ts]: Why did you have to make that change? It looks not related, smells like something bad is happening here.

File: src/main/features/offer/form/models/defendantResponse.ts [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+
+  constructor (option?: StatementType) {
+    this.option = option
+  }
+
+  static fromObject (value?: any): DefendantResponse {
+    if (!value) {
+      return value
+    }
+    // tslint:disable-next-line:no-console
```
Comment (timja-kainos) [inline src/main/features/offer/form/models/defendantResponse.ts]: umm
